src/ens/storage.c

```diff
@@ -0,0 +1,185 @@
+        if (rc <= 0) {
+            return rc;
+        }
+    }
+    k_mutex_unlock(&info_fs_lock);
+    return 0;
+}
+
+/**
+ * Save our current storage infromation to flash.
+ */
+int save_storage_information() {
+    k_mutex_lock(&info_fs_lock, K_FOREVER);
+    int rc = nvs_write(&info_fs, STORED_CONTACTS_INFO_ID, &record_information, sizeof(record_information));
+    if (rc <= 0) {
+        printk("Something went wrong after saving storage information.\n");
+    }
+    k_mutex_unlock(&info_fs_lock);
+    return rc;
+}
+
+record_sequence_number_t get_next_sequence_number() {
+    k_mutex_lock(&info_fs_lock, K_FOREVER);
+    if (record_information.count >= MAX_CONTACTS) {
+        record_information.oldest_contact = sn_increment(record_information.oldest_contact);
+    } else {
+        record_information.count++;
+    }
+    save_storage_information();
+    record_sequence_number_t next_sn = get_latest_sequence_number();
+    k_mutex_unlock(&info_fs_lock);
+    return next_sn;
+}
+
+int init_record_storage(void) {
+    int rc = 0;
+    struct flash_pages_info info;
+    // define the nvs file system
+    info_fs.offset = FLASH_AREA_OFFSET(storage);
+    rc =
+        flash_get_page_info_by_offs(device_get_binding(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL), info_fs.offset, &info);
+
+    if (rc) {
+        // Error during retrieval of page information
+        printk("Cannot retrieve page information (err %d)\n", rc);
+        return rc;
+    }
+    info_fs.sector_size = info.size;
+    info_fs.sector_count = SEC_COUNT;
+
+    rc = nvs_init(&info_fs, DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
+    k_mutex_init(&info_fs_lock);
+    if (rc) {
+        // Error during nvs_init
+        printk("Cannot init NVS (err %d)\n", rc);
+        return rc;
+    }
+
+    // Load the current storage information
+    rc = load_storage_information();
+    if (rc) {
+        printk("Cannot load storage information (err %d)\n", rc);
+        return rc;
+    }
+
+    printk("Currently %d contacts stored!\n", record_information.count);
+    printk("Space available: %d\n", FLASH_AREA_SIZE(storage));
+
+    rc = ens_fs_init(&ens_fs, FLASH_AREA_ID(ens_storage), sizeof(record_t));
+    if (rc) {
+        printk("Cannot init ens_fs (err %d)\n", rc);
+    }
+    return rc;
+}
+
+int load_record(record_t* dest, record_sequence_number_t sn) {
+    storage_id_t id = convert_sn_to_storage_id(sn);
+    int rc = ens_fs_read(&ens_fs, id, dest);
+    if (rc < 0) {
+        return rc;
+    }
+    return 0;
+}
+
+int add_record(record_t* src) {
+    int sectorsToDelete = 1;
+
+    k_mutex_lock(&info_fs_lock, K_FOREVER);
+
+    // Check, if next sn would be at start of page
+    record_sequence_number_t potential_next_sn = sn_increment(get_latest_sequence_number());
+    storage_id_t potential_next_id = convert_sn_to_storage_id(potential_next_sn);
+    if (((potential_next_id * ens_fs.entry_size) % ens_fs.sector_size) == 0) {
```

Comment:
We need to rework this: We simply try to write all the time (which might fail due to space constraints or invalid flash memory). If writing failed, we first try to make_space in the ens system (which also fails if we are in the middle of a page). If it succeeds, we simply try to write again. If clearing also fails, we increase, delete the occupied entry (to prevent future invalid data), increase the used sequence number, and try everything again.